design/design.md

```diff
@@ -0,0 +1,414 @@
+(or just changing interfaces in case of the simple lottery which itself is
+already a decentralized scheme), see section 4.1 and 4.2 in the paper.
+Additionally, the support for weights can be introduced in the decentralized
+schemes by having the simple lottery take into account weights, as done in
+Algorand's sortition (see section 5 in the paper).
+(Thus, the simple lottery in the decentralized weighted setting is equivalent
+to Algorand's sortition.)
+Finally, we can create centralized weighted ALBA schemes by remodeling the
+decentralized weighted schemes as centralized;
+this is currently the only known efficient way to implement weights.
+
+We will implement Telescope Bounded,
+as well as, the simple lottery in all four settings (centralized+decentralized,
+unweighted+weighted).
+The following is the full list of eight implemented schemes.
+* Centralized unweighted Telescope Bounded (section 3.2.2)
+* Decentralized unweighted Telescope Bounded (section 3.2.2 + 4.2)
+* Centralized+decentralized weighted Telescope Bounded (section 3.2.2 + 4.2 + 5)
+* Centralized+decentralized unweighted simple lottery (section 4.1)
+* Centralized+decentralized weighted simple lottery (section 4.1 + 5)
+
+#### Discussion
+We will not implement ALBA schemes (1) and (2) for the following reasons.
+The basic Telescope is way slower and offers almost no advantages over
+the other Telescope schemes.
+Telescope with Prehashing is only applicable when the number of honest
+prover's elements $n_p$ is large
+(about a million for typical parameters) and it offers no advantage
+over Telescope Bounded except for a tiny difference in the certificate size
+(the integer $v$).
+(As an optimization of Telescope Bounded, one can omit this integer when
+serializing a certificate if the number of allowed repetitions $r = 1$.)
+
+We will implement the simple lottery scheme; despite its large proof sizes
+it offers very fast proving time which might be important for some applications.
+Additionally, in the decentralized setting, it has the smallest communication
+complexity, so it's likely to be the top choice for applications where
+the certificate size is not important.
+
+We will implement schemes that seem redundant (e.g., unweighted schemes
+since the user could just use a weighted variant with each weight equal 1,
+or centralized schemes since the user could just use a decentralized variant
+directly) for two reasons:
+1. to provide a simpler interface for the user;
+2. because they might be more efficient now or when optimizations are
+implemented in the future; for instance, unweighted variants can be made
+more efficient due to no lottery and less complexity, and in the decentralized
+setting, an unweighted version could be more efficient due to not needing to
+sample the binomial distribution.
+
+We should be able to reuse most of the logic between different schemes anyway,
+so the overhead shouldn't be large.
+
+### Choice of hash function
```

Comment:
Likewise, you can be much more concise. 

I would not write at all about domain separation. We can say we shall implement it for Alba, and that users using Alba should also implement it on the application level. But that's pretty much it.

This is a design document for Alba, we should not explain too much about the rest, the part on using different hash functions and correlation is in particular: this is not in the scope of the document, and you actually may need to use different ones:a hash function for a hashmap has different requirements than one for zerocoin.

Reply:
I am not saying that domain separation is not important but that we should focus on our repository: we implemented domain separation as security good practice. We can recommend to also use it in the wider application, but it is not the goal of the document to say what should, or shouldn't be done, there. Hence I would keep it short.

Reply:
But this section directly affects the repository / library implementation: whether we allow the user to provide a custom hash function and what the interface should be. While the exact interface is not specified here, this sections sets guidance for the right interface.